### PR TITLE
Switch amd-ci to use MI300X runner.

### PR DESCRIPTION
**Pull request by saienduri**
This commit switches the amd-ci workflow to use MI300x gpu provided by AMD for testing coverage.

### CODE REVIEW

**Comment by ByronHsu** [comment]
The tests are failing. That was i what i meant by the need for an interactive env to debug. @tjtanaa can you help us look into the failure since you have AMD GPU access? Thanks!

**Comment by tjtanaa** [comment]
@ByronHsu 
# CI Setup
I am running my test on MI250 and I manage to run all the test successfully without experiencing the error that you are facing. 

I do not have clear details about the environment on your CI machine, but my suspicion is that the pytorch version and triton is installed might not be fully compatible.

Could you try to reinstall pytorch version and set triton to triton==3.0.0?

```
python3 -m pip uninstall -y torch torchvision \
            && python3 -m pip install --pre \
                torch==2.6.0.dev20241113+rocm6.2 \
                'setuptools-scm>=8' \
                torchvision==0.20.0.dev20241113+rocm6.2 \
                --extra-index-url https://download.pytorch.org/whl/nightly/rocm6.2
```
*Note:* You could also try using the Pytorch Stable version. However, I still use this nightly version because this is what the vLLM Dockerfile.rocm is using, which should be proven to be a stable one.

Install triton from upstream `python3 -m pip install triton==3.0.0`, to resolve the issue with triton's cache_manager https://github.com/triton-lang/triton/issues/5013 .

# Side note about Unit Test case
In the `test/transformers/test_cross_entropy.py::test_float32_internal`, the `num_warp` needs to be set to `16` on AMD hardware.

With this, all the tests should pass.

**Comment by ByronHsu** [comment]
@saienduri is it possible to run in docker?
@tjtanaa We are using triton 3.1.0 and torch 2.5.1. Is only 3.1.0 working on AMD?

image

**Comment by tjtanaa** [comment]
@ByronHsu 
I have tested `triton==3.0.0` and `triton==3.1.0` with `torch==2.6.0.dev20241113+rocm6.2`.
I don't quite understand the question `Is only 3.1.0 working on AMD?`. My hunch is that any triton version that is `>3.0.0` should work on AMD.

**Comment by saienduri** [comment]
Yes, we can run in a docker. Please provide which one we would like to run within along with `docker pull` and `docker run` commands

**Comment by ByronHsu** [comment]
Glad to know that triton >= 3.0.0 works on AMD. The screenshot is from https://github.com/linkedin/Liger-Kernel/actions/runs/12174881791/job/33957645620?pr=428. 

I love the docker idea. @tjtanaa can you create a docker image for liger and hand over to @saienduri ? Would be great if we can publish the image through our CI too.

**Comment by tjtanaa** [comment]
@saienduri @ByronHsu 
I have opened a PR https://github.com/linkedin/Liger-Kernel/pull/430 into your branch. Can you validate if the steps are working. I have validated the `Dockerfile.rocm` and the command added into the `.github/workflow/amd-ci.yml`. The only thing that is not tested is the whether the `amd-ci.yml` workflow is setup correctly.

The followings are some additional details which are also highlighted in the PR description.

**Details**
1. When running the `docker run` you might need to run with `sudo`. In this version, I have not included it. If you find problem using AMD GPU within the docker container. Please add back the `sudo`. E.g.
  ```bash
  sudo docker run \
     --network=host \
     --group-add=video \
     --ipc=host \
     --cap-add=SYS_PTRACE \
     --security-opt seccomp=unconfined \
     --device /dev/kfd \
     --device /dev/dri \
     liger-ci \
     /bin/bash -c "make test-convergence; make test"
  ```

2. The `transformers` version has been pinned to `4.46.3` as there is a API changes in the later version of `transformers`. E.g. I have tested with `transformers==4.47.0` and the following error occured.
  ```bash
  FAILED test/convergence/test_mini_models_multimodal.py::test_mini_model_multimodal[mini_qwen2_vl-32-0.0001-dtype0-1e-08-1e-05-0.005-1e-05-0.005-1e-05] - AttributeError: 'Qwen2VLConfig' object has no attribute 'video_token_id'. Did you mean: 'vision_token_id'?
  FAILED test/convergence/test_mini_models_multimodal.py::test_mini_model_multimodal[mini_qwen2_vl-32-0.0001-dtype1-0.001-0.01-0.1-0.01-0.01-0.01] - AttributeError: 'Qwen2VLConfig' object has no attribute 'video_token_id'. Did you mean: 'vision_token_id'?
  FAILED test/convergence/test_mini_models_no_logits.py::test_mini_model[mini_qwen2_vl-32-0.0001-dtype6-1e-08-1e-05-0.005-1e-05-0.005-1e-05] - AttributeError: 'Qwen2VLConfig' object has no attribute 'image_token_id'. Did you mean: 'pad_token_id'?
  FAILED test/convergence/test_mini_models_no_logits.py::test_mini_model[mini_qwen2_vl-32-0.0001-dtype7-0.001-0.01-0.1-0.01-0.01-0.01] - AttributeError: 'Qwen2VLConfig' object has no attribute 'image_token_id'. Did you mean: 'pad_token_id'?
  ```

**Comment by ByronHsu** [comment]
Thanks @tjtanaa! I have given you the maintainer access. You can now push to `saienduri/amd-ci` in the main repo. In the meantime, @tyler-romero can you look at the qwen2_vl issue?

**Comment by tjtanaa** [comment]
@ByronHsu 
All tests have passed through the following changes.

I have treated the `test_rms_norm::test_correctness` as flaky test for now by adding the decorator `@pytest.mark.flaky(reruns=3, reruns_delay=2)` to the test case.

I have also temporary added the condition to change the `num_warps` of the `test/transformers/test_cross_entropy.py::test_float32_internal()`

However, I do think that the flakiness requires more investigations later one. It is not flaky in natural, it consistently failed and the numerical values are always the same.

File: .github/workflows/amd-ci.yml [hunk 1]
```diff
@@ -53,12 +52,40 @@ jobs:
...
-    - name: Run tests
+    - name: Setup Dependencies
+      run: |
+        cp -r /opt/rocm/share/amd_smi ./
+        cd amd_smi
+        python -m pip install -e .
+        cd ..
+        python -m pip install pytest pytest-xdist pytest-rerunfailures pytest-flakefinder pytest-cpp
+        python -m pip uninstall -y torch torchvision
+        python -m pip install --pre \
```
Comment (ByronHsu) [inline .github/workflows/amd-ci.yml:77]: this is a bit nasty. Can we have a easy way to install amd dep? Like `pip install liger-kernel[amd]`

File: test/transformers/test_rms_norm.py [hunk 2]
```diff
@@ -74,6 +74,7 @@ def forward(self, x):
         return output.type_as(x)
 
 
+@pytest.mark.flaky(reruns=3, reruns_delay=2)
```
Comment (ByronHsu) [inline test/transformers/test_rms_norm.py:77]: will this count as "pass" after all rerun "fails"?

**Comment by ByronHsu** [comment]
Merging for now! Thanks for the great work @saienduri @tjtanaa! Let's resolve my comments in the following PRs

File: test/transformers/test_rms_norm.py (hunk 2 above)
Comment (tjtanaa) [reply test/transformers/test_rms_norm.py:77]: It will be counted as "FAILED".

File: .github/workflows/amd-ci.yml (hunk 1 above)
Comment (tjtanaa) [reply .github/workflows/amd-ci.yml:77]: Addressing this in PR https://github.com/linkedin/Liger-Kernel/pull/436.

**Pull request merged**